src/input/joystick.h

```diff
@@ -0,0 +1,101 @@
+#ifndef INPUT_JOYSTICK_H
+#define INPUT_JOYSTICK_H
+
+#include "input/mouse.h"
+
+#define JOYSTICK_MAX_NAME 64
+#define JOYSTICK_MAX_GUID 33
+#define JOYSTICK_MAX_AXIS 8
+#define JOYSTICK_MAX_BUTTONS 20
+#define JOYSTICK_MAX_TRACKBALLS 4
+#define JOYSTICK_MAX_HATS 4
+#define JOYSTICK_MAX_CONTROLLERS 6
+
+#define JOYSTICK_MAPPING_ELEMENTS_MAX 2
+
+typedef enum {
+    JOYSTICK_ELEMENT_NONE = 0,
+    JOYSTICK_ELEMENT_AXIS = 1,
+    JOYSTICK_ELEMENT_TRACKBALL = 2,
+    JOYSTICK_ELEMENT_BUTTON = 3,
+    JOYSTICK_ELEMENT_HAT = 4
+} joystick_element;
+
+typedef enum {
+    JOYSTICK_AXIS_POSITIVE = 0,
+    JOYSTICK_AXIS_NEGATIVE = 1
+} joystick_axis_position;
+
+typedef enum {
+    JOYSTICK_HAT_CENTERED = 0,
+    JOYSTICK_HAT_UP = 1,
+    JOYSTICK_HAT_LEFT = 2,
+    JOYSTICK_HAT_DOWN = 4,
+    JOYSTICK_HAT_RIGHT = 8
+} joystick_hat_position;
+
+typedef enum {
+    JOYSTICK_TRACKBALL_X_POSITIVE = 0,
+    JOYSTICK_TRACKBALL_X_NEGATIVE = 1,
+    JOYSTICK_TRACKBALL_Y_POSITIVE = 2,
+    JOYSTICK_TRACKBALL_Y_NEGATIVE = 3,
+} joystick_trackball_position;
+
+typedef enum {
+    MAPPING_ACTION_MOUSE_CURSOR_UP,
+    MAPPING_ACTION_MOUSE_CURSOR_LEFT,
+    MAPPING_ACTION_MOUSE_CURSOR_DOWN,
+    MAPPING_ACTION_MOUSE_CURSOR_RIGHT,
+    MAPPING_ACTION_FASTER_MOUSE_CURSOR_SPEED,
+    MAPPING_ACTION_SLOWER_MOUSE_CURSOR_SPEED,
+    MAPPING_ACTION_LEFT_MOUSE_BUTTON,
+    MAPPING_ACTION_RIGHT_MOUSE_BUTTON,
+    MAPPING_ACTION_SCROLL_WINDOW_UP,
+    MAPPING_ACTION_SCROLL_WINDOW_DOWN,
+    MAPPING_ACTION_SCROLL_MAP_UP,
+    MAPPING_ACTION_SCROLL_MAP_LEFT,
+    MAPPING_ACTION_SCROLL_MAP_DOWN,
+    MAPPING_ACTION_SCROLL_MAP_RIGHT,
+    MAPPING_ACTION_ROTATE_MAP_LEFT,
+    MAPPING_ACTION_ROTATE_MAP_RIGHT,
+    MAPPING_ACTION_INCREASE_GAME_SPEED,
+    MAPPING_ACTION_DECREASE_GAME_SPEED,
+    MAPPING_ACTION_TOGGLE_PAUSE,
+    MAPPING_ACTION_CYCLE_LEGION,
+    MAPPING_ACTION_SHOW_VIRTUAL_KEYBOARD,
+    MAPPING_ACTION_CYCLE_TOUCH_TYPE,
+    MAPPING_ACTION_RESET_MAPPING,
+    MAPPING_ACTION_MAX
+} mapping_action;
+
+typedef struct {
+    struct {
+        joystick_element type;
+        int id;
+        int position;
+    } element[JOYSTICK_MAPPING_ELEMENTS_MAX];
+} mapping_element;
+
+typedef struct {
+    int connected_joysticks;
+    char name[JOYSTICK_MAX_NAME];
+    char guid[JOYSTICK_MAX_GUID];
+    int total_axis;
+    int total_buttons;
+    int total_trackballs;
+    int total_hats;
+    mapping_element first_mapping[MAPPING_ACTION_MAX];
+    mapping_element second_mapping[MAPPING_ACTION_MAX];
+} joystick_model;
```

Comment:
Ideally I'd not expose this struct in the header, since it's only used internally in the corresponding `joystick.c`. See also my comments on the next two functions.

Reply:
I need to expose this in the header because both `platform/joystick.c` and `input/joystick.c` need full access to `joystick_model`.

`platform` because it fills the `model` with the values from SDL (and also from mapping, which could later be also obtained via SDL), and `input` needs to access the model information, mainly for mapping purposes.